calicoctl/commands/diags.go

```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return err
+	}
+
+	runDiags(arguments["--log-dir"].(string))
+
+	return nil
+}
+
+func runDiags(logDir string) {
+	cmds := []struct {
+		info string
+		cmd  string
+	}{
+		{"", "date"},
+		{"", "hostname"},
+		{"Dumping netstat", "netstat --all --numeric"},
+		{"Dumping routes (IPv4)", "ip -4 route"},
+		{"Dumping routes (IPv6)", "ip -6 route"},
+		{"Dumping interface info (IPv4)", "ip -4 addr"},
+		{"Dumping interface info (IPv6)", "ip -6 addr"},
+		{"Dumping iptables (IPv4)", "iptables-save"},
+		{"Dumping iptables (IPv6)", "ip6tables-save"},
+		{"Dumping ipsets", "ipset list"},
+		{"Dumping ipsets (container)", "docker run --privileged --net=host calico/node ipset list"},
+		{"Copying journal for calico-node.service", "journalctl -u calico-node.service --no-pager"},
+	}
+
+	// Make sure the command is run with super user priviladges
+	if os.Getuid() != 0 {
+		fmt.Println("Need super user privilages: Operation not permitted")
+		os.Exit(1)
+	}
+
+	fmt.Println("Collecting diagnostics")
+
+	// Create a temp directory in /tmp
+	tmpDir, err := ioutil.TempDir("/tmp", "tmp")
+	if err != nil {
+		log.Fatalf("Error creating temp directory to dump logs: %v\n", err)
+	}
+
+	fmt.Println("Using temp dir:", tmpDir)
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		log.Fatalf("Error changing directory to temp directory to dump logs: %v\n", err)
+	}
+
+	os.Mkdir("diagnostics", os.ModeDir)
+	diagsTmpDir := filepath.Join(tmpDir, "diagnostics")
+
+	for _, v := range cmds {
+		writeDiags(v.info, v.cmd, diagsTmpDir)
+	}
+
+	fmt.Println("Dumping felix stats")
+	err = exec.Command("pkill", "-SIGUSR1", "felix").Run()
+	if err != nil {
+		log.Printf("Error killing the felix process: %v\n", err)
+	}
+
+	fileInfo, err := os.Stat(logDir)
+
+	// Check if the logDir provided/default exists and is a directory
+	if fileInfo.IsDir() {
+		fmt.Println("Copying Calico logs")
+		cpCmd := exec.Command("cp", "-rf", logDir, diagsTmpDir)
+		err = cpCmd.Run()
+		if err != nil {
+			log.Fatalf("Error copying log files: %v\n", err)
+		}
+	} else {
+		fmt.Printf("No logs found in %s; skipping log copying", logDir)
+	}
+
+	// Get the current time and create a tar.gz file with the timestamp in the name
+	timeNow := time.Now()
+	y, mo, d := timeNow.Date()
+	h, m, s := timeNow.Clock()
+	tarFile := fmt.Sprintf("diags-%d%d%d_%d%d%d.tar.gz", d, mo, y, h, m, s)
+
+	err = exec.Command("tar", "-zcvf", tarFile, "diagnostics").Run()
+	if err != nil {
+		log.Printf("Error compressing the diagnostics: %v\n", err)
+	}
+
+	tarFilePath := filepath.Join(tmpDir, tarFile)
+
+	fmt.Printf("\nDiags saved to %s\n", tarFilePath)
+	fmt.Printf(`If required, you can upload the diagnostics bundle to a file sharing service 
+    such as transfer.sh using curl or similar.  For example:
+
+      curl --upload-file %s https://transfer.sh/%s\n`, tarFilePath, tarFilePath)
+
+}
+
+func writeDiags(info, cmd, dir string) {
+	fname := cmd
+
+	r := []regPair{
```

Comment:
I think lines 148-152 would be better as a separate function - this is basically converting the function name to a suitable filename.

In actual fact - do we even need this.  Since you've added a nice structure that links the user description and the command, you could easily add into that a filename - so you don't have to do this munging of the command.